onchain-program/src/processor.rs

```diff
@@ -0,0 +1,178 @@
+#[allow(unused_imports)]
+use {
+    crate::{
+        instruction::{parse_program_instruction, ProgramInstruction},
+        state,
+    },
+    solana_program::{
+        account_info::{next_account_info, AccountInfo},
+        clock::Clock,
+        entrypoint,
+        entrypoint::ProgramResult,
+        msg,
+        program::invoke_signed,
+        program_error::ProgramError,
+        program_utils::limited_deserialize,
+        pubkey::Pubkey,
+        rent::Rent,
+        system_instruction,
+        sysvar::Sysvar,
+    },
+};
+
+// Declare and export the program's entrypoint.
+entrypoint!(process_instruction);
+
+pub fn process_instruction(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    instruction_data: &[u8],
+) -> ProgramResult {
+    match parse_program_instruction(instruction_data)? {
+        ProgramInstruction::CreateBucket {
+            data,
+            size,
+            bump_seed,
+        } => Processor::create_bucket(program_id, accounts, data, size as usize, bump_seed),
+        ProgramInstruction::PutIntoBucket { data, offset } => {
+            Processor::put_into_bucket(program_id, accounts, data, offset as usize)
+        }
+    }
+}
+
+pub struct Processor;
+impl Processor {
+    fn create_bucket(
+        program_id: &Pubkey,
+        accounts: &[AccountInfo],
+        data: &[u8],
+        size: usize,
+        bump_seed: u8,
+    ) -> ProgramResult {
+        let account_info_iter = &mut accounts.iter();
+
+        let authority_account = next_account_info(account_info_iter)?;
+        let payer_account = next_account_info(account_info_iter)?;
+        let data_bucket_account = next_account_info(account_info_iter)?;
+        let system_program_account = next_account_info(account_info_iter)?;
+
+        let authority_key = *authority_account.signer_key().ok_or_else(|| {
```

Comment:
I think in this case because the payer and the authority is the same (from the CLI) it works, but I do not think you can have 2 accounts both being the signer unless it is some multisig account?

Reply:
Ok, I accounted for this better in the client and verified it. Having separate author & payer actually works. I think I got the original idea for this from the Solana `address-lookup-table` program (though in hindsight that's not 100% applicable due to being native program).